### PR TITLE
add servant-queryparam

### PR DESCRIPTION
**Pull request by deemp**
Hi! I'd like to add here the `servant-queryparam` (ex - [named-servant](https://hackage.haskell.org/package/named-servant)) family of libraries.
These libraries add support for `QueryParam`s as records or named parameters.

What I changed:
- added a type family for modification of fields of query parameter records. Here's an [example](https://github.com/deemp/servant/blob/ceb5a5a282491a3b9fee0eba00bf3b7f91588098/servant-named/servant-named-server/example/Main.hs).
- renamed the packages: `named-servant` -> `servant-queryparam-core`, etc.

I first used a modified version of the `symbols` package that I provided using a Nix flake. 
I later moved the type family from https://github.com/kcsongor/symbols/pull/4 to `servant-named-core`. 
So, technically, a flake is not necessary to build `servant-named`, but it can improve the build time.

## Example

`servant-named-server` provides an example of constructing an OpenAPI3 specification for an API with query parameters.

To run the  example, run `nix develop -c cabal run servant-named-server:exe:example`.

The output should look like

```json
---
Query parameters without prefixes
---

{
    "components": {},
    "info": {
        "title": "",
        "version": ""
    },
    "openapi": "3.0.0",
    "paths": {
        "/get": {
            "get": {
                "parameters": [
                    {
                        "in": "query",
                        "name": "user",
                        "required": false,
                        "schema": {
                            "type": "string"
                        }
                    }
                ],
                "responses": {
                    "200": {
                        "content": {
                            "application/json;charset=utf-8": {
                                "schema": {
                                    "maximum": 9223372036854775807,
                                    "minimum": -9223372036854775808,
                                    "type": "integer"
                                }
                            }
                        },
                        "description": ""
                    },
                    "400": {
                        "description": "Invalid `user`"
                    }
                }
            }
        }
    }
}

---
Query parameters with prefixes
---

{
    "components": {},
    "info": {
        "title": "",
        "version": ""
    },
    "openapi": "3.0.0",
    "paths": {
        "/get": {
            "get": {
                "parameters": [
                    {
                        "in": "query",
                        "name": "_get_user",
                        "required": false,
                        "schema": {
                            "type": "string"
                        }
                    }
                ],
                "responses": {
                    "200": {
                        "content": {
                            "application/json;charset=utf-8": {
                                "schema": {
                                    "maximum": 9223372036854775807,
                                    "minimum": -9223372036854775808,
                                    "type": "integer"
                                }
                            }
                        },
                        "description": ""
                    },
                    "400": {
                        "description": "Invalid `_get_user`"
                    }
                }
            }
        }
    }
}
```

### CODE REVIEW

File: servant-named/servant-named-client/src/Servant/Client/Record.hs [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- | -- | This module just exports orphan instances to make named-servant
```
Comment (maksbotan) [inline servant-named/servant-named-client/src/Servant/Client/Record.hs:14]:
```suggestion
-- | This module just exports orphan instances to make named-servant
```

also, `named-servant`?

File: servant-named/servant-named-core/src/Servant/Named.hs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- |
+-- This module uses the named package to match names with parameters. For example, this api:
+-- 
+-- @
+-- type API = "users" :> (QueryParam "category" Category :>
```
Comment (maksbotan) [inline servant-named/servant-named-core/src/Servant/Named.hs:15]: where is the `(` closed?

File: servant-named/servant-named-core/src/Servant/Named.hs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+-- @
+-- type API = "users" :> (OptionalQueryParam "category" Category :>
+--                        NamedQueryParam "sort_by" SortBy :>
+--                        NamedQueryFlag "with_schema" :>
+--                        NamedQueryParams "filters" Filter :>
+--                        Get '[JSON] User
+-- @
+-- 
+-- The servant-named-client and servant-named-server will create
+-- functions that use the `named` package to match the names with the
```
Comment (maksbotan) [inline servant-named/servant-named-core/src/Servant/Named.hs:33]:
```suggestion
-- functions that use the @named@ package to match the names with the
```
afair haddock can't link to packages

File: servant-named/servant-named-core/src/Servant/Named.hs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+-- | Like `QueryParam'`, but instead of extracting a type @a@, it
+-- extracts a named type @`NamedF` f a sym@, where the name
+-- corresponds to the query parameter string.
+data NamedQueryParam' (mods :: [Type]) (sym :: Symbol) (a :: Type)
+
+unarg :: NamedF f a name -> f a
+unarg (ArgF a) = a
+
+-- | type family to rewrite a named queryparam to a regular
```
Comment (maksbotan) [inline servant-named/servant-named-core/src/Servant/Named.hs:60]:
```suggestion
-- | Type family to rewrite a named queryparam to a regular
```

File: servant-named/servant-named-core/src/Servant/Named.hs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+instance
+  (KnownSymbol sym, ToHttpApiData v, HasLink sub) =>
+  HasLink (NamedQueryParams sym v :> sub)
+  where
+  type MkLink (NamedQueryParams sym v :> sub) a = sym :? [v] -> MkLink sub a
+  toLink toA _ l (ArgF params) =
+    toLink toA (Proxy :: Proxy (QueryParams sym v :> sub)) l $
+      fromMaybe [] params
+
+-- | Like `QueryFlag, but extracts a named type @named `:?` Bool@
```
Comment (maksbotan) [inline servant-named/servant-named-core/src/Servant/Named.hs:110]:
```suggestion
-- | Like `QueryFlag`, but extracts a named type @named `:?` Bool@
```

File: servant-named/servant-named-core/src/Servant/Record.hs [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+module Servant.Record (RecordParam, UnRecordParam) where
+
+import Data.Kind (Type)
+import Data.Proxy
+import GHC.Generics
+import GHC.TypeLits (KnownSymbol, Symbol)
+import Servant.API
+import Servant.Symbols (Exp, Modify)
+
+-- import Servant.Symbols (Exp, Modify)
```
Comment (maksbotan) [inline servant-named/servant-named-core/src/Servant/Record.hs:22]: ?

File: servant-named/servant-named-core/src/Servant/Record.hs [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+--   { category :: Maybe Category
+--   , sort_by :: Sortby
+--   , with_schema :: Bool
+--   , filters :: [Filter]
+--   }
+--
+-- type API = "users" :> RecordParam SomeLabel UserParams :> Get '[JSON] User
+-- 
+-- 'SomeLabel' will be used to evaluate the field modifiers using 'Modify'
+-- @
```
Comment (maksbotan) [inline servant-named/servant-named-core/src/Servant/Record.hs:49]:
```suggestion
-- type API = "users" :> RecordParam SomeLabel UserParams :> Get '[JSON] User
-- @ 
--
-- 'SomeLabel' will be used to evaluate the field modifiers using 'Modify'
```

File: servant-named/servant-named-server/src/Servant/Server/Named.hs [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    hoistServerWithContext (Proxy :: Proxy api) pc nt . s
+
+  route Proxy context subserver =
+    route (Proxy :: Proxy (QueryParam' mods sym a :> api)) context $
+      fmap
+        ( \f x ->
+            case sbool :: SBool (FoldRequired mods) of
+              STrue -> case sbool :: SBool (FoldLenient mods) of
+                STrue -> f (Arg x)
+                SFalse -> f (Arg x)
```
Comment (maksbotan) [inline servant-named/servant-named-server/src/Servant/Server/Named.hs:71]: is the nested case necessary? both branches look the same

File: servant-named/servant-named-core/src/Servant/Symbols.hs [hunk 9]
```diff
@@ -0,0 +1,143 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- |
+-- This module uses ideas from the Chapter 10 of the book @Thinking with Types@ by Sandy Maguire.
```
Comment (maksbotan) [inline servant-named/servant-named-core/src/Servant/Symbols.hs:9]: shouldn't this be in a separate package on hackage? maybe there is one already?

**Review by maksbotan — COMMENTED** [review]
there are sevaral warnings in the CI log about the new packages, like redundant imports, unused exports, `StarIsType` and so on

**Comment by maksbotan** [comment]
@tchoutri what do you think about the inclusion of this package in servant repo?

File: servant-named/servant-named-server/src/Servant/Server/Named.hs (hunk 8 above)
Comment (deemp) [reply servant-named/servant-named-server/src/Servant/Server/Named.hs:71]: Yes, there goes some singleton magic.

```hs
x :: If
  (FoldRequired' 'False mods)
  (If (FoldLenient' 'False mods) (Either Text a) a)
  (Maybe (If (FoldLenient' 'False mods) (Either Text a) a))
```

File: servant-named/servant-named-core/src/Servant/Symbols.hs (hunk 9 above)
Comment (deemp) [reply servant-named/servant-named-core/src/Servant/Symbols.hs]: Indeed, there was a way to use the `first-class-families` package directly. Still, some useful expressions were missing there so they're grouped into a module.

**Comment by deemp** [comment]
I decided to make a separate repo https://github.com/deemp/servant-queryparam

**Pull request closed without merging**